### PR TITLE
Combine probs

### PR DESCRIPTION
**Pull request by anirudh-challa**
This PR is related to https://github.com/triagemd/keras-eval/issues/55

### CODE REVIEW

File: keras_eval/eval.py [hunk 1]
```diff
@@ -102,7 +105,8 @@ def _get_complete_image_paths(self, filenames):
         return image_paths
 
     def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix=False,
-                 save_confusion_matrix_path=None):
+                 save_confusion_matrix_path=None, combine_training_classes=False, combine_classes_dict_dir=None,
```
[the remark this answers is not in the document]
Comment (anirudh-challa) [reply keras_eval/eval.py]: I will check for errors in the dictionary but I don't think it is necessary for the user to provide an `id`, we assume that the order they write the file in is the order they want. Also, I think it will be more readable with `sub_class` name as when anyone looks at the `.json` will understand which classes are being grouped together.

File: keras_eval/eval.py [hunk 2]
```diff
@@ -53,6 +54,8 @@ def __init__(self, **options):
         self.model_specs = []
         self.probabilities = None
         self.labels = None
+        self.training_concepts = None
```
[the remark this answers is not in the document]
Comment (anirudh-challa) [reply keras_eval/eval.py]: 1. I forgot to remove `self.training_concepts `, will do that. 
2. I'd prefer to name it something else other than `virtual_concepts` as virtual is an overloaded term already with a different meaning. Right now I am calling them `combine_classes_dict`.

File: keras_eval/eval.py [hunk 3]
```diff
@@ -137,6 +137,34 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
             # Obtain labels to show on the metrics results
             self.concept_labels = utils.get_concept_items(self.concepts, key='label')
 
+            if combine_training_classes:
```
[the remark this answers is not in the document]
Comment (anirudh-challa) [reply keras_eval/eval.py]: Yes, I was thinking of that as well. Will do.  I will also need to pass the training class dictionary to that function.

**Comment by adriaromero** [comment]
After some discussions, we decided to just ask the user a **single dictionary** that provides (1) the training class ordering and (2) the class combination (merging the old `dictionary.json` and `combine_class_dic.json` in a single one). The proposed dictionary structure is the following:

```
[
  {
    "class_index": 0,
    "class_name": "cat",
    "test_class": "domestic"
  },
  {
    "class_index": 1,
    "class_name": "dog",
    "test_class": "domestic"
  },
  {
    "class_index": 2,
    "class_name": "goose",
    "test_class": "water"
  },
  {
    "class_index": 3,
    "class_name": "turtle",
    "test_class": "water"
  },
  {
    "class_index": 4,
    "class_name": "elephant",
    "test_class": "wild"
  }
]
```

File: keras_eval/eval.py [hunk 4]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
 
         return self.probabilities, self.labels
 
+    def evaluate_combination_concepts(self):
+        '''
+        Sets probabilities if model concepts are to be combined in test
```
Comment (adriaromero) [inline keras_eval/eval.py]: Can we rephrase it to include what the function uses (`self.concept_dictionary`, `self.probabilities`)?

Something like:
> Combine probabilities following `self.concept_dictionary` criteria and rewrite them in `self.probabilities`

File: keras_eval/eval.py [hunk 5]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
...
+        '''
+
+        if utils.compare_group_test_concepts(self.concept_labels, self.concept_dictionary) and utils.check_concept_unique(self.concept_dictionary):
```
Comment (adriaromero) [inline keras_eval/eval.py]: Why not combining `utils.compare_group_test_concepts` and `utils.check_concept_unique` to be a single function `utils.check_concept_dictionary`? 

I don't see other uses `compare_group_test_concepts` and `check_concept_unique` can have individually.

File: keras_eval/eval.py [hunk 6]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
...
+            for clas in self.concept_dictionary:
```
Comment (adriaromero) [inline keras_eval/eval.py]: typo error: ~`clas`~ --> `class`

Why not calling `class` --> `concept` ?

File: keras_eval/eval.py [hunk 7]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
...
+
+                if clas['group'] in self.group_id_dict.keys():
+                    self.group_id_dict[clas['group']].append(clas['class_index'])
+                else:
+                    self.group_id_dict[clas['group']] = [clas['class_index']]
+
+            self.combined_probs = [[[0.0, ] * len(self.concept_labels)] * len(self.probabilities[0])]
+            self.combined_probs = np.array(self.combined_probs)
+            id = 0
+            for concept_label in self.concept_labels:
```
Comment (adriaromero) [inline keras_eval/eval.py]: You can avoid incrementing the `id` by doing:
```
for idx, concept_label in enumerate(self.concept_labels):
```

and `idx` will increment for each `self.concept_labels` item.

File: keras_eval/eval.py [hunk 8]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
...
+                col_nums = self.group_id_dict[concept_label]
```
Comment (adriaromero) [inline keras_eval/eval.py]: I'm not really fan of abbreviations 🤔 . Can we find a better name for `col_nums`?

File: keras_eval/eval.py [hunk 9]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
...
+                for col_num in col_nums:
+                    self.combined_probs[0][:, id] += self.probabilities[0][:, col_num]
+                id += 1
+            self.probabilities = self.combined_probs
+
+        else:
+            raise ValueError('There are issues with the .json file provided')
```
Comment (adriaromero) [inline keras_eval/eval.py]: You don't need to add the `else` statement since if any error occurs in the `if`, it will raise before arriving to this point.

File: keras_eval/utils.py [hunk 10]
```diff
@@ -113,6 +113,67 @@ def get_concept_items(concepts, key):
...
+        with open(dictionary_path, 'r') as dictionary_file:
+            dict = json.load(dictionary_file)
+    else:
+        raise ValueError('Error: invalid dictionary path' + str(dictionary_path))
+    return dict
+
+
+def create_training_json(train_dir, output_json_file):
+    '''
+            Checks if evaluation concepts are unique
```
Comment (adriaromero) [inline keras_eval/utils.py]: Check indentation here. 

It should be:
```
'''
Checks if evaluation concepts are unique
Args:
    train_dir: The location where you have the training directory
    output_json_file: The output file name and path e.g.: ./dictionary.json
Returns:
    True, if there are no repeat concepts, else raises error
'''
```

File: keras_eval/utils.py [hunk 11]
```diff
@@ -113,6 +113,67 @@ def get_concept_items(concepts, key):
...
+            Args:
+                train_dir: The location where you have the training directory
+                output_json_file: The output file name and path e.g.: ./dictionary.json
+
+            Returns:
+                True, if there are no repeat concepts, else raises error
+    '''
+    dict = []
```
Comment (adriaromero) [inline keras_eval/utils.py]: `dict` is a Python predefined string. Choose a different one!

File: keras_eval/utils.py [hunk 12]
```diff
@@ -113,6 +113,67 @@ def get_concept_items(concepts, key):
...
+    train_concepts = get_default_concepts(train_dir)
+    for i in range(len(train_concepts)):
```
Comment (adriaromero) [inline keras_eval/utils.py]: Since `i` will store in `class_index`, why not rename it to `idx`?

File: keras_eval/utils.py [hunk 13]
```diff
@@ -113,6 +113,67 @@ def get_concept_items(concepts, key):
...
+        dict.append({"class_index": i, "class_name": train_concepts[i]["label"], "group": train_concepts[i]["label"]})
+    with open(output_json_file, 'w') as file_obj:
+        json.dump(dict, file_obj)
+
+
+def check_concept_unique(concept_dict):
+    '''
+        Checks if evaluation concepts are unique
```
Comment (adriaromero) [inline keras_eval/utils.py]: Check code indentation.

File: keras_eval/utils.py [hunk 14]
```diff
@@ -113,6 +113,67 @@ def get_concept_items(concepts, key):
...
+            raise ValueError("Concept has been repeated:", concept_dict_item['class_name'])
+        else:
+            concept_class_name_dict[concept_dict_item['class_name']] = 1
+
+    return True
+
+
+def compare_group_test_concepts(test_concepts_list, concept_dict):
+    '''
+        Checks if concept dictionary has the groups as the test concepts
```
Comment (adriaromero) [inline keras_eval/utils.py]: Check code indentation.

File: keras_eval/utils.py [hunk 15]
```diff
@@ -113,6 +113,67 @@ def get_concept_items(concepts, key):
...
+            concept_dict: Dictionary that contains class_id, train_concepts and groups
+        Returns:
+            True, if there are no repeat concepts, else raises error
+    '''
+    concept_class_name_dict = {}
+    for concept_dict_item in concept_dict:
+        if concept_dict_item['class_name'] in concept_class_name_dict:
+            raise ValueError("Concept has been repeated:", concept_dict_item['class_name'])
+        else:
+            concept_class_name_dict[concept_dict_item['class_name']] = 1
```
Comment (adriaromero) [inline keras_eval/utils.py]: Maybe a best way to avoid storing this `1`?

File: script/fetch-fixtures [hunk 16]
```diff
@@ -1,9 +1,9 @@
 #!/usr/bin/env bash
 set -e
 (
-mkdir -p tmp/fixtures/models/mobilenet_1/ tmp/fixtures/models/mobilenet_2/
+mkdir -p tmp/fixtures/models/test_1/mobilenet_1/ tmp/fixtures/models/test_1/mobilenet_2/
```
Comment (adriaromero) [inline script/fetch-fixtures]: rename `test_1` folder to `ensemble`

File: script/fetch-fixtures [hunk 17]
```diff
@@ -1,9 +1,9 @@
...
 
-cd tmp/fixtures/models/mobilenet_1/
+cd tmp/fixtures/models/test_1/mobilenet_1/
```
Comment (adriaromero) [inline script/fetch-fixtures]: Same: `test_1` --> `ensemble`

File: script/fetch-fixtures [hunk 18]
```diff
@@ -12,4 +12,12 @@ cd ../mobilenet_2/
 if [ ! -f mobilenet_v1.h5 ]; then
     wget $MODEL_URL
 fi
-)
\ No newline at end of file
+)
+
+mkdir -p tmp/fixtures/models/test_2/mobilenet_3/
```
Comment (adriaromero) [inline script/fetch-fixtures]: rename `test_2` --> `single` and remove `mobilenet_3` level.

File: script/fetch-fixtures [hunk 19]
```diff
@@ -12,4 +12,12 @@ cd ../mobilenet_2/
...
+
+cd tmp/fixtures/models/test_2/mobilenet_3/
```
Comment (adriaromero) [inline script/fetch-fixtures]: Replace ~`tmp/fixtures/models/test_2/mobilenet_3/`~ with `tmp/fixtures/models/single/`

File: tests/files/animals/combine_class_dic.json [hunk 20]
```diff
@@ -0,0 +1,13 @@
+[
```
Comment (adriaromero) [inline tests/files/animals/combine_class_dic.json]: Duplicated file. Please, remove.

File: tests/files/animals/test/dictionary.json [hunk 21]
```diff
@@ -0,0 +1,17 @@
+[
```
Comment (adriaromero) [inline tests/files/animals/test/dictionary.json]: Not used anymore. Please, remove.

File: tests/files/animals/test/manifest.json [hunk 22]
```diff
@@ -0,0 +1,213 @@
+[
```
Comment (adriaromero) [inline tests/files/animals/test/manifest.json]: File not used. Please, remove.

File: tests/test_eval.py [hunk 23]
```diff
@@ -31,12 +36,12 @@ def evaluator_mobilenet():
              'target_size': [224, 224, 3]
              }
 
-    with open(os.path.abspath('tmp/fixtures/models/mobilenet_1/model_spec.json'), 'w') as outfile:
+    with open(os.path.abspath('tmp/fixtures/models/test_1/mobilenet_1/model_spec.json'), 'w') as outfile:
```
Comment (adriaromero) [inline tests/test_eval.py]: Rename `test_1` to be `ensemble`

File: tests/test_eval.py [hunk 24]
```diff
@@ -31,12 +36,12 @@ def evaluator_mobilenet():
...
         json.dump(specs, outfile)
 
     return Evaluator(
         batch_size=1,
-        model_path='tmp/fixtures/models/mobilenet_1/mobilenet_v1.h5'
+        model_path='tmp/fixtures/models/test_1/mobilenet_1/mobilenet_v1.h5'
```
Comment (adriaromero) [inline tests/test_eval.py]: Rename `test_1` to be `ensemble`

File: tests/test_eval.py [hunk 25]
```diff
@@ -49,19 +54,38 @@ def evaluator_ensemble_mobilenet():
              'target_size': [224, 224, 3]
              }
 
-    with open(os.path.abspath('tmp/fixtures/models/mobilenet_1/model_spec.json'), 'w') as outfile:
+    with open(os.path.abspath('tmp/fixtures/models/test_1/mobilenet_1/model_spec.json'), 'w') as outfile:
```
Comment (adriaromero) [inline tests/test_eval.py]: Rename `test_1` to be `ensemble`

File: tests/test_eval.py [hunk 26]
```diff
@@ -49,19 +54,38 @@ def evaluator_ensemble_mobilenet():
...
         json.dump(specs, outfile)
 
-    with open(os.path.abspath('tmp/fixtures/models/mobilenet_2/model_spec.json'), 'w') as outfile:
+    with open(os.path.abspath('tmp/fixtures/models/test_1/mobilenet_2/model_spec.json'), 'w') as outfile:
```
Comment (adriaromero) [inline tests/test_eval.py]: Rename `test_1` to be `ensemble`

File: tests/test_eval.py [hunk 27]
```diff
@@ -49,19 +54,38 @@ def evaluator_ensemble_mobilenet():
...
         json.dump(specs, outfile)
 
     return Evaluator(
-        ensemble_models_dir='tmp/fixtures/models/',
+        ensemble_models_dir='tmp/fixtures/models/test_1/',
```
Comment (adriaromero) [inline tests/test_eval.py]: Rename `test_1` to be `ensemble`

File: tests/test_eval.py [hunk 28]
```diff
@@ -49,19 +54,38 @@ def evaluator_ensemble_mobilenet():
...
+@pytest.fixture('function')
+def evaluator_mobilenet_class_combine():
+    specs = {'klass': 'keras.applications.mobilenet.MobileNet',
+             'name': 'mobilenet_v1',
+             'preprocess_args': [123.99345370133717, 116.22568321228027, 99.73750913143158],
+             'preprocess_func': 'mean_subtraction',
+             'target_size': [299, 299, 3]
+             }
+
+    with open(os.path.abspath('tmp/fixtures/models/test_2/mobilenet_3/model_spec.json'), 'w') as outfile:
```
Comment (adriaromero) [inline tests/test_eval.py]: Rename `test_2/mobilenet_3` to be `single`

File: tests/test_eval.py [hunk 29]
```diff
@@ -49,19 +54,38 @@ def evaluator_ensemble_mobilenet():
...
+        json.dump(specs, outfile)
+
+    return Evaluator(
+        batch_size=1,
+        model_path='tmp/fixtures/models/test_2/mobilenet_3/animals_combine_classes.hdf5',
```
Comment (adriaromero) [inline tests/test_eval.py]: Rename `test_2/mobilenet_3` to be `single`

File: tests/test_utils.py [hunk 30]
```diff
@@ -41,10 +46,17 @@ def test_round_list():
     assert utils.round_list(input_list, decimals=6) == [0.666667, 0.333333]
 
 
+def test_read_dictionary(training_dict_file):
+    dict = utils.read_dictionary(training_dict_file)
```
Comment (adriaromero) [inline tests/test_utils.py]: `dict` is a predefined Python string. Use a different one 🙏

File: tests/test_utils.py [hunk 31]
```diff
@@ -41,10 +46,17 @@ def test_round_list():
...
+    expected = 5
+    actual = len(dict)
+    assert actual == expected
+
+
 def test_load_model():
     custom_objects = {'relu6': mobilenet.layers.ReLU(6, name='relu6'), "tf": tf}
-    model_path = 'tmp/fixtures/models/mobilenet_1/mobilenet_v1.h5'
-    model_spec_path = 'tmp/fixtures/models/mobilenet_2/model_spec.json'
+    model_path = 'tmp/fixtures/models/test_1/mobilenet_1/mobilenet_v1.h5'
```
Comment (adriaromero) [inline tests/test_utils.py]: rename `test_1` to be `ensemble`

File: tests/test_utils.py [hunk 32]
```diff
@@ -41,10 +46,17 @@ def test_round_list():
...
+    model_spec_path = 'tmp/fixtures/models/test_1/mobilenet_2/model_spec.json'
```
Comment (adriaromero) [inline tests/test_utils.py]: rename `test_1` to be `ensemble`

File: tests/test_utils.py [hunk 33]
```diff
@@ -56,7 +68,7 @@ def test_load_model():
 
 
 def test_load_model_ensemble():
-    ensemble_dir = 'tmp/fixtures/models'
+    ensemble_dir = 'tmp/fixtures/models/test_1/'
```
Comment (adriaromero) [inline tests/test_utils.py]: rename `test_1` to be `ensemble`

File: tests/test_utils.py [hunk 34]
```diff
@@ -125,6 +137,33 @@ def test_default_concepts(test_dataset_path):
...
+    dict_path = './tests/files/dict.json'
+    utils.create_training_json(test_dataset_path, dict_path)
+    actual = os.path.isfile(dict_path)
+    expected = True
+    assert actual == expected
+
+
+def test_compare_concept_dictionaries():
+    concept_lst = ['dog', 'elephant']
+    dict = [{'group': 'dog'}, {'group': 'cat'}, {'group': 'elephant'}]
```
Comment (adriaromero) [inline tests/test_utils.py]: `dict` is a predefined Python string. Use a different one 🙏

File: tests/test_utils.py [hunk 35]
```diff
@@ -125,6 +137,33 @@ def test_default_concepts(test_dataset_path):
...
+    with pytest.raises(ValueError) as exception:
+        utils.compare_group_test_concepts(concept_lst, dict)
+    expected = 'The concept dictionary groups do not match the class labels'
+    actual = str(exception).split('ValueError: ')[1]
+    assert actual == expected
+
+
+def test_check_concept_unique():
+    dict = [{'class_name': 'cat'}, {'class_name': 'dog'}, {'class_name': 'cat'}]
```
Comment (adriaromero) [inline tests/test_utils.py]: `dict` is a predefined Python string. Use a different one 🙏

File: tests/test_utils.py [hunk 36]
```diff
@@ -168,7 +207,7 @@ def test_show_results():
 
 
 def test_ensemble_models(test_image_path, model_spec_mobilenet):
-    ensemble_dir = 'tmp/fixtures/models'
+    ensemble_dir = 'tmp/fixtures/models/test_1/'
```
Comment (adriaromero) [inline tests/test_utils.py]: rename `test_1` to be `ensemble`

File: keras_eval/utils.py (hunk 15 above)
Comment (anirudh-challa) [reply keras_eval/utils.py]: Could not find a better way of doing in O(1).

File: keras_eval/eval.py [hunk 37]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
 
         return self.probabilities, self.labels
 
+    def compute_inference_probabilities(self, concept_dictionary, probabilities):
+        '''
+        Sets probabilities if model concepts are to be combined in test
```
Comment (adriaromero) [inline keras_eval/eval.py]: Can we make this description more clear? 

Something like:
> Apply probability inference on concepts defined in `concept_dictionary` and save them in `self.probabilities`

File: keras_eval/eval.py [hunk 38]
```diff
@@ -145,6 +149,32 @@ def evaluate(self, data_dir=None, top_k=1, filter_indices=None, confusion_matrix
...
+
+        if utils.compare_group_test_concepts(self.concept_labels, concept_dictionary) and utils.check_concept_unique(concept_dictionary):
+            for concept in concept_dictionary:
+
+                if concept['group'] in self.group_id_dict.keys():
+                    self.group_id_dict[concept['group']].append(concept['class_index'])
+                else:
+                    self.group_id_dict[concept['group']] = [concept['class_index']]
+
+            self.combined_probs = [[[0.0, ] * len(self.concept_labels)] * len(probabilities[0])]
```
Comment (adriaromero) [inline keras_eval/eval.py]: If we rename the function to `compute_inference_probabilities`, may we rename `combined_probs` as `inferred_probs` to be coherent?

File: keras_eval/utils.py [hunk 39]
```diff
@@ -113,6 +113,67 @@ def get_concept_items(concepts, key):
...
+
+    return True
+
+
+def compare_group_test_concepts(test_concepts_list, concept_dict):
+    '''
+    Checks if concept dictionary has the groups as the test concepts
+    Args:
+        test_concepts_list: List of lables of test concepts
```
Comment (adriaromero) [inline keras_eval/utils.py]: In order to avoid the 2x`of`s, one option could be: 
> List of labels corresponding to the test concepts

File: tests/test_eval.py [hunk 40]
```diff
@@ -49,19 +54,38 @@ def evaluator_ensemble_mobilenet():
...
     return Evaluator(
-        ensemble_models_dir='tmp/fixtures/models/',
+        ensemble_models_dir='tmp/fixtures/models/ensemble/',
         combination_mode='arithmetic',
         batch_size=1
     )
 
 
+@pytest.fixture('function')
+def evaluator_mobilenet_class_combine():
```
Comment (adriaromero) [inline tests/test_eval.py]: If we're going to start using the "inference" naming, may we change all the "combine" words with "inference"? 🤔

File: keras_eval/eval.py (hunk 37 above)
Comment (anirudh-challa) [reply keras_eval/eval.py]: I am changing it to "Combines probabilities based on key "group" in concept_dictionary and saves the values in self.probabilities"

File: README.md [hunk 41]
```diff
@@ -61,6 +63,82 @@ ensemble_models_dir = '/model_folder'
...
+    "group": "00001_water"
+  },
+  {
+    "class_index": 4,
+    "class_name": "00004_elephant",
+    "group": "00002_wild"
+  }
+]
+```
+Here are the results for this scenario:
```
Comment (jsalbert) [inline README.md]: looks good to me!

File: README.md [hunk 42]
```diff
@@ -61,6 +63,82 @@ ensemble_models_dir = '/model_folder'
...
+[test_set_0] class_0 or class_1
+[test_set_1] class_2 or class_3
+```
+
+So the way we combine probabilities is as below:
+```
+probability(test_set_0) =  probability(class_0) + probability(class_1)
+probability(test_set_1) = probability(class_2) + probability(class_3)
+```
+We would want the users to give us the mapping between the training and testing dictionary as a `.json` file. Given below is the format we expect:
```
Comment (adriaromero) [inline README.md]: More formal explanations needed here (note this is not a PR, it's a ReadMe):
> For this purpose, the mapping between the training and testing dictionary must be provided as a `.json` file with the following format:

File: README.md (hunk 41 above)
Comment (adriaromero) [reply README.md]: Add:
> Confusion matrix for inference evaluation:

File: README.md [hunk 43]
```diff
@@ -61,6 +63,82 @@ ensemble_models_dir = '/model_folder'
...
+  }
+]
+```
+Here are the results for this scenario:
+![Confusion_matrix](https://github.com/triagemd/keras-eval/blob/combine_probs/figs/confusion_matrix_coarse.png)
+Below are the average results for this scenario:
+
+model | accuracy | precision | f1_score | number_of_samples	| number_of_classes
+-- | -- | -- | -- | -- | -- 
+0	animals_combine_classes.hdf5 | 0.733 | 0.841	| 0.729	| 15	| 3
```
Comment (adriaromero) [inline README.md]: Remove the `0` and extra blank spaces from `model` cell.

File: README.md [hunk 44]
```diff
@@ -61,6 +63,82 @@ ensemble_models_dir = '/model_folder'
...
+
+So in the example above the group gives us the mapping between a single concept during training and the concepts which we would want to evaluate on in test.
```
Comment (adriaromero) [inline README.md]: This is redundant. I think it's already implicit in the title description above.

File: README.md [hunk 45]
```diff
@@ -61,6 +63,83 @@ ensemble_models_dir = '/model_folder'
...
+]
+```
+Results for class consolidated evaluation:
+![Confusion_matrix](https://github.com/triagemd/keras-eval/blob/combine_probs/figs/confusion_matrix_coarse.png)
+
+model | accuracy | precision | f1_score | number_of_samples	| number_of_classes
+-- | -- | -- | -- | -- | -- 
+animals_combine_classes.hdf5 | 0.733 | 0.841	| 0.729	| 15	| 3 
 
 You can specify all the following options.
```
Comment (adriaromero) [inline README.md]: We should indicate the new `Evaluator` argument `concept_dictionary_path` here.

File: README.md [hunk 46]
```diff
@@ -61,6 +63,83 @@ ensemble_models_dir = '/model_folder'
...
+```
+Regular evaluation scenario:
+```
+[class_0]
+[class_1]
+[class_2]
+[class_3]
+```
+Results for regular evaluation:
+![Confusion_matrix](https://github.com/triagemd/keras-eval/blob/combine_probs/figs/confusion_matrix_granular.png)
```
Comment (adriaromero) [inline README.md]: We'll need to rename `combine_probs` to `master`, otherwise the image won't be shown in the Master branch.
